File: COM Automation/InProcess/MATLAB/TestCOMTSMaster.m

APP_CAN                       = 0;
APP_LIN                       = 1;
BUS_UNKNOWN_TYPE              = 0;
TS_TCP_DEVICE                 = 1;
TS_USB_DEVICE                 = 3;
XL_USB_DEVICE                 = 2;
lfdmACKOff                    = 1;
lfdmNormal                    = 0;
lfdmRestricted                = 2;
lfdtCAN                       = 0;
lfdtISOCAN                    = 1;
lfdtNonISOCAN                 = 2;
LVL_ERROR                     = 1;
LVL_WARNING                   = 2;
LVL_OK                        = 3;
LVL_HINT                      = 4;
LVL_INFO                      = 5;
LVL_VERBOSE                   = 6;

APP_NAME = "TestCOMTSMaster";
USE_XL_VIRTUAL_DEVICE = false;
USE_XL_VN1630_DEVICE = false;
USE_XL_VN1640_DEVICE = false;
USE_TS_VIRTUAL_DEVICE = true;
USE_TS_CANMINI_DEVICE = false;

fprintf('Script started\n');

% retrieve TSMaster application management
app = actxserver('comTSMaster.TSApplication');

% set current application name
app.set_current_application(APP_NAME);
fprintf('current application is %s\n', app.get_current_application());

% get all application in a list separated by ";"
fprintf("application list is: %s\n", app.get_application_list());

% set CAN channel count to 2
app.set_can_channel_count(2);

% set LIN channel count to 2
app.set_lin_channel_count(0);

% read CAN channel count
fprintf("CAN channel count = %d\n", app.get_can_channel_count());

% read LIN channel count
fprintf("LIN channel count = %d\n", app.get_lin_channel_count());

% delete mapping of application CAN channel 1 and 2
app.del_mapping(APP_CAN, 0);
app.del_mapping(APP_CAN, 1);

% map application channel 1 to hw index 0, channel 0 of Vector virtual channel
r.FAppName = APP_NAME;
r.FAppChannelIndex = 0;
r.FAppChannelType = APP_CAN;
if USE_XL_VIRTUAL_DEVICE
	r.FHWDeviceType = XL_USB_DEVICE;
	r.FHWDeviceSubType = 1;
elseif USE_XL_VN1630_DEVICE
	r.FHWDeviceType = XL_USB_DEVICE;
	r.FHWDeviceSubType = 57;
elseif USE_XL_VN1640_DEVICE
	r.FHWDeviceType = XL_USB_DEVICE;
	r.FHWDeviceSubType = 59;
elseif USE_TS_VIRTUAL_DEVICE
	r.FHWDeviceType = TS_TCP_DEVICE;
	r.FHWDeviceSubType = -1;
elseif USE_TS_CANMINI_DEVICE
	r.FHWDeviceType = TS_USB_DEVICE;
	r.FHWDeviceSubType = 3;
end
r.FHWIndex = 0;
r.FHWChannelIndex = 0;
r.FHWDeviceName = "Virtual";
r.FMappingDisabled = false;
app.set_mapping_verbose(r.FAppChannelIndex, r.FAppChannelType, ...
    r.FHWDeviceType, r.FHWIndex, r.FHWChannelIndex, r.FHWDeviceSubType, ...
    r.FHWDeviceName, r.FMappingDisabled);

% map application channel 2 to hw index 0, channel 1 of Vector virtual channel
r.FAppChannelIndex = 1;
r.FHWChannelIndex = 1;
app.set_mapping_verbose(r.FAppChannelIndex, r.FAppChannelType, ...
    r.FHWDeviceType, r.FHWIndex, r.FHWChannelIndex, r.FHWDeviceSubType, ...
    r.FHWDeviceName, r.FMappingDisabled);

% print application CAN channel mappings
[r.FHWDeviceType, r.FHWIndex, r.FHWChannelIndex, r.FHWDeviceSubType, ...
    r.FHWDeviceName, r.FMappingDisabled] = app.get_mapping_verbose(APP_CAN, 0);
fprintf("Application CAN channel 1 mapping to: %s of Channel %d\n", r.FHWDeviceName, r.FHWChannelIndex);
[r.FHWDeviceType, r.FHWIndex, r.FHWChannelIndex, r.FHWDeviceSubType, ...
    r.FHWDeviceName, r.FMappingDisabled] = app.get_mapping_verbose(APP_CAN, 1);
fprintf("Application CAN channel 1 mapping to: %s of Channel %d\n", r.FHWDeviceName, r.FHWChannelIndex);

% configure CAN channel baudrates before connection: Arb = 500K, Data = 2M
app.configure_baudrate_canfd(0, 500, 2000, lfdtISOCAN, lfdmNormal, true);
app.configure_baudrate_canfd(1, 500, 2000, lfdtISOCAN, lfdmNormal, true);

% connect application
app.log('connecting application...', LVL_HINT);
app.connect();

% retrieve application timestamp
fprintf("Current timestamp(s) is %f\n", app.get_timestamp() / 1000000.0);

% get communication object
com = app.TSCOM();
registerevent(com, {'OnRxCAN', @OnRxCAN});
registerevent(com, {'OnRxCANFD', @OnRxCANFD});
registerevent(com, {'OnRxLIN', @OnRxLIN});

% enable CAN message reception event, select only one entry from CAN and CAN FD events
com.enable_event_can(true);
com.enable_event_canfd(false);

% MATLAB will convert any single column matrix to a 1 dimensional array when passed to a COM object.
feature('COM_SafeArraySingleDim', 0);
feature('COM_PassSafeArrayByRef', 0);

% display function prototypes
invoke(com)
events(com)

% init a TCAN record for CAN message transmission
c.FIdxChn = 0;
c.FIsExtendedId = false;
c.FIsRemote = false;
c.FIdentifier = hex2dec('123');
c.FDLC = 8;
c.FDatas = '0, 5, 20, 3, 159, 52, 134, 1';

% transmit this classical CAN frame asynchrnously
com.transmit_can_async_verbose(c.FIdxChn, c.FIsRemote, c.FIsExtendedId, ...
    c.FDLC, c.FIdentifier, c.FDatas);

% retrieve application timestamp
app.wait(100);
fprintf("Current timestamp(s) is %f\n", app.get_timestamp() / 1000000.0);
 
% enable CAN FD message reception event, select only one entry from CAN and CAN FD events
com.enable_event_can(false);
com.enable_event_canfd(true);

% init a TCANFD record for CAN FD message transmission
cFD.FIdxChn = 0;
cFD.FIsExtendedId = 0;
cFD.FIsEDL = true;
cFD.FIsBRS = true;
cFD.FIsESI = false;
cFD.FIdentifier = hex2dec('345');
cFD.FDLC = 15;
cFD.FDatas = '1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64';
 
% transmit this CAN FD frame asynchrnously
com.transmit_canfd_async_verbose(cFD.FIdxChn, cFD.FIsExtendedId, ...
    cFD.FIsEDL, cFD.FIsBRS, cFD.FIsESI, ...
    cFD.FDLC, cFD.FIdentifier, cFD.FDatas);
 
% retrieve application timestamp
app.wait(100);
fprintf("Current timestamp(s) is %f\n", app.get_timestamp() / 1000000.0);
 
% send periodically
com.add_cyclic_msg_can_verbose(c.FIdxChn, c.FIsExtendedId, c.FIdentifier, ...
    c.FDLC, c.FDatas, 100);
app.wait(1000);
com.delete_cyclic_msg_can_verbose(c.FIdxChn, c.FIsExtendedId, c.FIdentifier);
com.add_cyclic_msg_canfd_verbose(cFD.FIdxChn, cFD.FIsExtendedId, cFD.FIdentifier, ...
    cFD.FDLC, cFD.FDatas, 100);
app.wait(1000);
 
% update message data
app.log('updating message data and cycle time...', LVL_OK);
c.FDatas = '8, 7, 6, 5, 4, 3, 2, 1';
cFD.FDatas = '8, 7, 6, 5, 4, 3, 2, 1, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 8, 7, 6, 5, 4, 3, 2, 1';
com.add_cyclic_msg_can_verbose(c.FIdxChn, c.FIsExtendedId, c.FIdentifier, ...
    c.FDLC, c.FDatas, 500);
com.add_cyclic_msg_canfd_verbose(cFD.FIdxChn, cFD.FIsExtendedId, cFD.FIdentifier, ...
    cFD.FDLC, cFD.FDatas, 200);
app.wait(1000);
 
% stop periodic messages
com.delete_cyclic_msgs();
app.wait(1000);

% unregister all events
unregisterallevents(com);

% disconnect application
app.disconnect();
app.log('closing application...', LVL_INFO);
 
% finalize library
fprintf("Script finalized\n");

% Arg. No.
% 1     Object Name
% 2     Event ID
% 3     Start of Event argument list
% end-2 End of event argument list (argument N)
% end-1 event structure
% end   event name
function OnRxCAN(varargin)
    % 1 * 13, varargin = {1¡Á1 Interface.comTSMaster_ITSCOM}    {[201]}    {' '}    {[0]}    {[0]}    {[0]}    {[0]}    {'}    {[273]}    {[224432829]}    {1¡Á8 uint8}    {1¡Á1 struct}    {'OnRxCAN'}
    s = ['CAN message received, Channel = %d, Id = 0x%3x, DLC = %d, Time = %f, Data = '  repmat('%2x ', 1, 8) '\n'];
    fprintf(s, varargin{4}, varargin{9}, varargin{8}, single(varargin{10})/1000000.0, varargin{11});
end

function OnRxCANFD(varargin)
    % 1 * 15, varargin = {1¡Á1 Interface.comTSMaster_ITSCOM}    {[202]}    {' '}    {[0]}    {[0]}    {[0]}    {[1]}    {[0]}    {[0]}    {'}    {[273]}    {[217054228]}    {1¡Á64 uint8}    {1¡Á1 struct}    {'OnRxCANFD'}
    s = ['CAN FD message received, Channel = %d, Id = 0x%3x, DLC = %d, Time = %f, Data = '  repmat('%2x ', 1, 64) '\n'];
    fprintf(s, varargin{4}, varargin{11}, varargin{10}, single(varargin{12})/1000000.0, varargin{13});
end

function OnRxLIN(varargin)
    % 1 * 12, varargin = {1¡Á1 Interface.comTSMaster_ITSCOM}    {[201]}    {' '}    {[0]}    {[0]}    {[0]}    {'}    {[273]}    {[224432829]}    {1¡Á8 uint8}    {1¡Á1 struct}    {'OnRxLIN'}
    s = ['LIN message received, Channel = %d, Id = 0x%3x, DLC = %d, Time = %f, Data = '  repmat('%2x ', 1, 8) '\n'];
    fprintf(s, varargin{4}, varargin{8}, varargin{7}, single(varargin{9})/1000000.0, varargin{10});
end

File: COM Automation/InProcess/Python/TestCOMTSMaster.py
import win32com.client
import pythoncom
import win32api
import time
from win32com.client import VARIANT

class constants:
	APP_CAN                       = 0          # from enum TTSAppChannelType
	APP_LIN                       = 1          # from enum TTSAppChannelType
	BUS_UNKNOWN_TYPE              = 0          # from enum TTSBusToolDeviceType
	TS_TCP_DEVICE                 = 1          # from enum TTSBusToolDeviceType
	TS_USB_DEVICE                 = 3          # from enum TTSBusToolDeviceType
	XL_USB_DEVICE                 = 2          # from enum TTSBusToolDeviceType
	lfdmACKOff                    = 1          # from enum TTSCANFDControllerMode
	lfdmNormal                    = 0          # from enum TTSCANFDControllerMode
	lfdmRestricted                = 2          # from enum TTSCANFDControllerMode
	lfdtCAN                       = 0          # from enum TTSCANFDControllerType
	lfdtISOCAN                    = 1          # from enum TTSCANFDControllerType
	lfdtNonISOCAN                 = 2          # from enum TTSCANFDControllerType
	LVL_ERROR                     = 1
	LVL_WARNING                   = 2
	LVL_OK                        = 3
	LVL_HINT                      = 4
	LVL_INFO                      = 5
	LVL_VERBOSE                   = 6
	
APP_NAME = "TestCOMTSMaster"
USE_XL_VIRTUAL_DEVICE = False
USE_XL_VN1630_DEVICE = False
USE_XL_VN1640_DEVICE = False
USE_TS_VIRTUAL_DEVICE = True
USE_TS_CANMINI_DEVICE = False

class TSMasterEvents:
	def OnRxCAN(self, AIdxChn, AIsTx, AIsRemote, AIsExtended, AIsError, ADLC, AIdentifier, ATimeUs, ADatas):
		print("CAN message received:", AIdxChn, AIdentifier, ADLC, ATimeUs/1000000.0, [ADatas[0], ADatas[1], ADatas[2], ADatas[3], ADatas[4], ADatas[5], ADatas[6], ADatas[7]])
	def OnRxCANFD(self, AIdxChn, AIsTx, AIsExtended, AIsError, AIsEDL, AIsBRS, AIsESI, ADLC, AIdentifier, ATimeUs, ADatas):
		if AIsEDL:
			print("CAN FD message received:", AIdxChn, AIdentifier, ADLC, ATimeUs/1000000.0, [ADatas[0], ADatas[1], ADatas[2], ADatas[3], ADatas[4], ADatas[5], ADatas[6], ADatas[7], ADatas[8], ADatas[9], ADatas[10], ADatas[11], ADatas[12], ADatas[13], ADatas[14], ADatas[15], ADatas[16], ADatas[17], ADatas[18], ADatas[19], ADatas[20], ADatas[21], ADatas[22], ADatas[23], ADatas[24], ADatas[25], ADatas[26], ADatas[27], ADatas[28], ADatas[29], ADatas[30], ADatas[31], ADatas[32], ADatas[33], ADatas[34], ADatas[35], ADatas[36], ADatas[37], ADatas[38], ADatas[39], ADatas[40], ADatas[41], ADatas[42], ADatas[43], ADatas[44], ADatas[45], ADatas[46], ADatas[47], ADatas[48], ADatas[49], ADatas[50], ADatas[51], ADatas[52], ADatas[53], ADatas[54], ADatas[55], ADatas[56], ADatas[57], ADatas[58], ADatas[59], ADatas[60], ADatas[61], ADatas[62], ADatas[63]])
		else:
			print("CAN message received:", AIdxChn, AIdentifier, ADLC, ATimeUs/1000000.0, [ADatas[0], ADatas[1], ADatas[2], ADatas[3], ADatas[4], ADatas[5], ADatas[6], ADatas[7]])
	def OnRxLIN(self, AIdxChn, AIsTx, AIsError, ADLC, AIdentifier, AChecksum, ATimeUs, ADatas):
		print("LIN message received:")
	
print("Script started")

# retrieve TSMaster application management
#pythoncom.CoInitialize() # enable multithread
app = win32com.client.Dispatch("comTSMaster.TSApplication")

# set current application name
app.set_current_application(APP_NAME)
print("current application is:", app.get_current_application())

# get all application in a list separated by ";"
print("application list is:", app.get_application_list())

# set CAN channel count to 2
app.set_can_channel_count(2)

# set LIN channel count to 2
app.set_lin_channel_count(0)

# read CAN channel count
print("CAN channel count =", app.get_can_channel_count())

# read LIN channel count
print("LIN channel count =", app.get_lin_channel_count())

# delete mapping of application CAN channel 1 and 2
app.del_mapping(constants.APP_CAN, 0)
app.del_mapping(constants.APP_CAN, 1)

# map application channel 1 to hw index 0, channel 0 of Vector virtual channel
r = win32com.client.Record("TTSMapping", app)
r.FAppName = APP_NAME
r.FAppChannelIndex = 0
r.FAppChannelType = constants.APP_CAN
if USE_XL_VIRTUAL_DEVICE:
	r.FHWDeviceType = constants.XL_USB_DEVICE
	r.FHWDeviceSubType = 1
if USE_XL_VN1630_DEVICE:
	r.FHWDeviceType = constants.XL_USB_DEVICE
	r.FHWDeviceSubType = 57
if USE_XL_VN1640_DEVICE:
	r.FHWDeviceType = constants.XL_USB_DEVICE
	r.FHWDeviceSubType = 59
if USE_TS_VIRTUAL_DEVICE:
	r.FHWDeviceType = constants.TS_TCP_DEVICE
	r.FHWDeviceSubType = -1
if USE_TS_CANMINI_DEVICE:
	r.FHWDeviceType = constants.TS_USB_DEVICE
	r.FHWDeviceSubType = 3
r.FHWIndex = 0
r.FHWChannelIndex = 0
r.FHWDeviceName = "Virtual"
r.FMappingDisabled = False
app.set_mapping(r)

# map application channel 2 to hw index 0, channel 1 of Vector virtual channel
r.FAppChannelIndex = 1
r.FHWChannelIndex = 1
app.set_mapping(r)

# print application CAN channel mappings
print("Application CAN channel 1 mapping to:", app.get_mapping(constants.APP_CAN, 0))
print("Application CAN channel 2 mapping to:", app.get_mapping(constants.APP_CAN, 1))

# configure CAN channel baudrates before connection: Arb = 500K, Data = 2M
app.configure_baudrate_canfd(0, 500, 2000, constants.lfdtISOCAN, constants.lfdmNormal, True)
app.configure_baudrate_canfd(1, 500, 2000, constants.lfdtISOCAN, constants.lfdmNormal, True)

# connect application
app.log('connecting application...', constants.LVL_HINT)
app.connect()

# retrieve application timestamp
print("Current timestamp(s) is", app.get_timestamp() / 1000000.0)

# get communication object
com = app.TSCOM()
win32com.client.WithEvents(com, TSMasterEvents)

# enable CAN message reception event, select only one entry from CAN and CAN FD events
com.enable_bus_statistics(True);
com.enable_event_can(True)
com.enable_event_canfd(False)

# init a TCAN record for CAN message transmission
c = win32com.client.Record("TCAN", app)
c.FIdxChn = 0
c.FIsExtendedId = 0
c.FIsRemote = False
c.FIdentifier = 0x123
c.FDLC = 8
c.FDatas = VARIANT(pythoncom.VT_ARRAY | pythoncom.VT_I1, [1, 2, 3, 4, 5, 6, 7, 8])

# transmit this classical CAN frame asynchrnously
com.transmit_can_async(c)

# retrieve application timestamp
app.wait(100)
print("Current timestamp(s) is", app.get_timestamp() / 1000000.0)

# enable CAN FD message reception event, select only one entry from CAN and CAN FD events
com.enable_event_can(False)
com.enable_event_canfd(True)

# init a TCANFD record for CAN FD message transmission
cFD = win32com.client.Record("TCANFD", app)
cFD.FIdxChn = 0
cFD.FIsExtendedId = 0
cFD.FIsEDL = True
cFD.FIsBRS = True
cFD.FIsESI = False
cFD.FIdentifier = 0x345
cFD.FDLC = 15
cFD.FDatas = VARIANT(pythoncom.VT_ARRAY | pythoncom.VT_I1, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64])

# transmit this CAN FD frame asynchrnously
com.transmit_canfd_async(cFD)

# retrieve application timestamp
app.wait(100)
print("Current timestamp(s) is", app.get_timestamp() / 1000000.0)

# send periodically
com.add_cyclic_msg_can(c, 100)
com.add_cyclic_msg_canfd(cFD, 100)
app.wait(1000)

# update message data
app.log('updating message data...', constants.LVL_OK)
c.FDatas = VARIANT(pythoncom.VT_ARRAY | pythoncom.VT_I1, [8, 7, 6, 5, 4, 3, 2, 1])
cFD.FDatas = VARIANT(pythoncom.VT_ARRAY | pythoncom.VT_I1, [8, 7, 6, 5, 4, 3, 2, 1, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 8, 7, 6, 5, 4, 3, 2, 1])
com.add_cyclic_msg_can(c, 100)
com.add_cyclic_msg_canfd(cFD, 100)
app.wait(1000)

# stop periodic messages
com.delete_cyclic_msgs()
app.wait(1000)

r, s = app.get_system_var_generic('Application.Connected');
if r:
   print('Application.Connected = ' + s)
   
# disconnect application
app.log('closing application...', constants.LVL_INFO)
app.disconnect()

r, s = app.get_system_var_generic('Application.Connected');
if r:
   print('Application.Connected = ' + s)

# finalize library
app = None
print("Script finalized")
#input("press any key to continue...")
